### DISCUSSION TITLE
[Proposal] const structs

### DISCUSSION THREAD

**Discussion by JustNrik** [dotnet/csharplang#2878] [Language Ideas]
Basically, a "const struct" is a `readonly struct` that can be represented as a compile-time constant, like `System.Decimal` and `System.DateTime` (in VB)

How would this work and what are the benefit of this?

This constant structures would be pretty much static readonly fields under the hood, with the difference that, for performance purpose, their values will be replaced with an actual constant when possible.

Example:

```cs
public const struct Color
{
    private readonly uint _value;
    public const Color(uint value) // this ctor is exclusively for the constant declaration
    {
        if (value > 0xFFFFFF)
            throw new InvalidOperationException(); // To tell the compiler this isn't allowed

        _value = value;
    }
}
```
Usage:
```cs
public const Color Cyan = 0x00FFFF; // Valid because at compile-time it's lower than 0xFFFFFF and so it can be safely casted to uint and call the proper constructor.
```
This would also allow to allow c# to have datetime constant like vb does like:
```cs
public const DateTime Today = "9/10/2019"; 
```

Some possible rules:

1. const structs are implicitly readonly structs, and hence all their members must be readonly.
2. const structs must define at least 1 const ctor, which must take exactly 1 parameter that must be a constant value (not user-made), like string or any other primitive.
3. const structs are explicitly convertible to the constant they represent, and they will be replaced after compilation to their respective constant when possible and safe. For example, `Console.Write(Color.Red);`, here it's not safe to use the constant value it represents so the object is used instead. But if you wanted to do `var mangenta = new Color(Color.Blue | Color.Red);` and no `|` operator is defined, it's safe to replace with the uint value because it's valid in that context and there's no custom operator provided, hence it's replaced with the constants. It can also be replaced with the constant it represents if there's no valid possibility for the struct itself, like doing bit operations with Color and Color doesn't define any of these operators, since the operators can't take the Color itself as argument, they take the constant value it represents instead.

**Comment by HaloFour** [comment]
What would the following compile into?

```cs
public const DateTime Today = "9/10/2019"; 
```

More importantly, what would the following compile into?

```cs
public DateTime GetToday() => Today;
```

**Reply by semimono** [reply]
C# can already detect a const integer being assigned to a non-const value. This is allowed:
```public const int MyValue = 13 - (-1 < 0 ? -(-1) : -1);```
And this isn't allowed:
```public const int MyValue = 13 - System.Math.Abs(-1);```
Other than being wrapped in a "named" function call, this is no different. If it can tag an expression as const vs. non-const, it should be able to tag a simple function as const (and anything too complex or obviously not constant as non-const).

Switch statements require const values (makes good sense). I have a struct that wraps a byte with two values encoded into it (two enums each with just a few possible values, hence it fits) and a constructor that abstracts how it's encoded into a byte. I can calculate out the constant values by hand and hard code them in, but what if the values for the enums change? And, how readable is the resulting switch statement with "random" numbers rather than the pair of enums they actually represent? I don't necessarily need a const struct, but a function that can be substituted in place of a const value (a function which is equivalent to a const expression). A constructor would do the job, and of course switching on the struct that wraps the byte value would be nicer than switching on the byte, though generic support of switching on structs, even if const structs were allowed, would be an additional bit of work.

I understand this feature could be very large in scope, possibly far too large to ever implement given how rare the use case is, but that does not make it any less legitimate of a request.

**Reply by HaloFour** [reply]
> Other than being wrapped in a "named" function call, this is no different.

The difference is that the compiler knows what `(-1 < 0 ? -(-1) : -1)` will do, how long it will take to evaluate and how safe it is to execute.  That becomes significantly more complicated once the logic is behind a method call.  It's completely impossible if that method happens to exist in a reference assembly where the method body doesn't exist.  IMO, the majority of these use cases are better solved by using `readonly` instead of `const` and evaluating the expression once at runtime.  I'd rather see `switch` and pattern matching made more flexible and support non-const operands.

> but that does not make it any less legitimate of a request.

I don't think that the request is illegitimate.  Const expressions and compile-time evaluation would be an interesting addition.  But it's important to shake out any potential "gotchas" and compare the proposal against existing features or other proposals as a part of the process.

**Reply by semimono** [reply]
Ok makes sense. I think if a switch statement could be run against runtime values then it would make this functionality only important for performance.

**Comment by yaakov-h** [comment]
This bears similarities to #688 in it's goals, but seems to veer wildly into unpredictable territory like Halo demonstrated.

**Comment by JustNrik** [comment]
> 
> 
> What would the following compile into?
> 
> ```cs
> public const DateTime Today = "9/10/2019"; 
> ```
> 
> More importantly, what would the following compile into?
> 
> ```cs
> public DateTime GetToday() => Today;
> ```

https://sharplab.io/#v2:DYLgbgRgPgCgrhYBLAxgAgMLAIYGdeYCwAUGmWvIqpgPYB2uALmgCo0Am2AnmgIIEARbIwCmaALxoAxAEYADAHp5CgExyZATikkAonXaYc+IA===

**Comment by HaloFour** [comment]
@JustNrik 

I know how C# and VB.NET embed non-const values into const-like fields such as `decimal` and `DateTime`.  That doesn't tell me how you propose this `const struct` syntax to be implemented.

**Comment by Izumemori** [comment]
As far as I can tell you're going for something like compile time macros here, which would definitely be interesting to see and I like the idea behind this feature. But I feel like an implementation through structs is clumsy. I feel like something like this would need new syntax to be added, and then there's the question if this feature is that useful that is warrants having syntax added, maybe even compiler changes depending on how it's going to be implemented.

**Comment by sgf** [comment]
const structs is very usefully.
such as do something like make a Int24,UInt24.....
we cant define the custome struct as a const value.

**Comment by sgf** [comment]
> What would the following compile into?
> 
> ```cs
> public const DateTime Today = "9/10/2019"; 
> ```
> 
> More importantly, what would the following compile into?
> 
> ```cs
> public DateTime GetToday() => Today;
> ```

const is a compile-time behavior, which is the premise of using const.
so its value must be determined at compile time.
 Therefore, the variable names you give are misleading.
This is essentially the same as
```cs 
public  const string Today = "" 9/10/2019 "";
```

**Comment by HaloFour** [comment]
@sgf 

> so its value must be determined at compile time.

Indeed, the question is what is actually emitted.  A `String` has a native representation that can be embedded as bytes in the assembly.  A `DateTime` does not.  You have to execute code in order to create a `DateTime` of any particular value, and the code that needs to be executed differs for quite literally every `struct`.  That's the problem that you'd need to solve: how to turn that `const` expression into bytes that can be embedded into the assembly that can then be interpreted as a value of that `struct` at runtime.

**Comment by HaloFour** [comment]
For example, VB.NET fakes `DateTime` constants by creating a `readonly` field of `DateTime` and using an attribute to specify a "constant" value in ticks:

```vb
Const d As Date = #01/01/2019#
```

is really

```cs
[DateTimeConstant(636818976000000000L)]
private static readonly DateTime d = new DateTime(636818976000000000L);
```

And each compiler has to understand how to interpret that annotated field as a constant and not a field.

C# does the same with `decimal`.

The runtime doesn't understand these pseudo-constants and you can't use them in attributes or other places where the runtime itself needs to know how to hydrate the value.

So to make this work in a more general purpose manner you'd need a way to represent the struct as something that can be embedded in one or more other primitive values.  Perhaps that can be done with blittable structs.

But is there really much value to it?  For any `static readonly` it would only be initialized once and can be referenced directly.  And the value can change per execution, unlike compile-time constants which are embedded directly into the assembly.

**Comment by inf9144** [comment]
Could we than have switch/case over those const structs? That would be really awesome, because my own enums are for various reasons static fields with struct values. I could make those const then. Could we use them in attributes too?

**Comment by sgf** [comment]
> But is there really much value to it? For any static readonly it would only be initialized once and can be referenced directly. And the value can change per execution, unlike compile-time constants which are embedded directly into the assembly.

The value lies in improving the C # application startup speed. Perform precalculations under appropriate conditions. The const modifier makes some structs that will never change after startup more semantic.

**Comment by HaloFour** [comment]
@sgf 

> The value lies in improving the C # application startup speed. Perform precalculations under appropriate conditions. The const modifier makes some structs that will never change after startup more semantic.

How is that any different to a `readonly static`?

**Comment by inf9144** [comment]
@HaloFour see my comment - using the structs and const fields where only compile time consts can be used. As example attributes or switch/case.

**Comment by sgf** [comment]
> @sgf
> 
> > The value lies in improving the C # application startup speed. Perform precalculations under appropriate conditions. The const modifier makes some structs that will never change after startup more semantic.
> 
> How is that any different to a `readonly static`?

i think 'readonly static' maybe  need static constructor when the application startup.

**Comment by HaloFour** [comment]
@inf9144 

That requires answering all of my previous comments

Currently faked "const" structs are implemented with deep knowledge of the compiler.  That's how C# allows `decimal` to be used as a `const` and how VB.NET does the same with `Date`.  That syntax only results in a `readonly static` with a runtime evaluation based on custom metadata embedded in attributes.  You'd need a generic way to express that, and a generic way to evaluate const expressions. 
 That metadata would need to be agreed upon by the ecosystem.  This is why C# can't consume optional `Date` parameters from VB.NET, C# doesn't understand how to evaluate and instantiate a `DateTime` for the parameter from the custom attribute.

And this still doesn't solve for attributes where the metadata needs to be embedded directly into the assembly in a way that the runtime itself needs to understand.

So this proposal requires:

1.  Some generic way to express the bits of any struct in metadata.  This may be easier for blittable structs.
1.  Constant expressions supported by the language.
1.  Runtime support to understand how to hydrate the struct from the bits embedded in metadata.

Of the 3 I think constant expressions are the tallest order.  This issue has a list of other issues/discussions on that subject: https://github.com/dotnet/csharplang/issues/2379

**Comment by sgf** [comment]
> Could we than have switch/case over those const structs? That would be really awesome, because my own enums are for various reasons static fields with struct values. I could make those const then. Could we use them in attributes too?

i think when const struct supported,Yes we can.

**Comment by CodeHatchling** [comment]
If I recall my experience working with C, structs are simply a way to organize (and minimize carpal tunnel injury rates while typing) declarations and operations on groups of primitives, and to simplify pointer arithmetic. Think along the lines of "instances namespaces"; or perhaps primitives as files, and structs as folders. It is a simple yet powerful idea that mainly serves as a human convienience for performing bulk operations with less code and less mess, and I sometimes wonder if C# muddied that up somewhere along the way.

In theory, even if there was some weird technical issue that prevented this, code like this:

**struct GlobalPosition { public float Latitude, Longitude; } // All fields must be public
const GlobalPosition YourHouse = // Cannot use a constructor
{
    Latitude = 39.98996554109341f, 
    Longitude = -86.0862521615366f
};
...
Console.WriteLine($"{YourHouse.Latitude}{YourHouse.Longitude}");**

Could just be replaced with this before compile time:

**const float YourHouseLatitude = 39.98996554109341f;
const float YourHouseLongitude = -86.0862521615366f;
...
Console.WriteLine($"{YourHouseLatitude}{YourHouseLongitude}");**

And you'd just have uglier code.

**Reply by svick** [reply]
That's not "in theory", even without `const struct`, [both versions of that code produce the same machine code today](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQ65IGwAJG4DC6A3urhfjvgCy4BKApgOYCuANmAE4DKwXrAMbAAFAEpylMmkqzcAcXYB7AEZh2ABSUQAlsB1KAdrgCaS1lwAS5iI1wBeXIcYB3cbmlyvuADJh9wKwAJnaOAMwAnAB0EQAcERHYAKxJNEgADBFhaQBmADSS3pQ+Rsx6waG4ALSx2FHptQhJyNhISWHY2DmFcgC+ANw9slgRIgAkAEQkZhbWrLZRfgEVvdPmVjaMi6XlIb0TYoMylL09PVh4cHSERhDAAGrcOmAq7IwQ4j2e3oK3wLg5ZT+UzrOa2Ja7SqRGLxRIpNKZbJIHJHIq/Qx3AFA/4zDbzRglQxlQIhBzVWr1RrNHBtDpdVHeEbjKa4sEE/yQ1aszaE4krA4MiinNDC9B3ATCBTKNSabR6AzGEj4MJYpTAiEkxh5Xw7TX9XDCoA).

**Reply by CodeHatchling** [reply]
Thanks, that certainly helps prove my point. :D
